packages/react/src/components/FocusTrapZone/FocusTrapZone.tsx

```diff
@@ -97,284 +92,230 @@ export const FocusTrapZone: React.FunctionComponent<IFocusTrapZoneProps> & {
+    // be sure to use the latest values here
+    if (disabled) {
+      return;
+    }
+    if (internalState.focusStackId === FocusTrapZone.focusStack!.slice(-1)[0]) {
+      const targetElement = ev.target as HTMLElement | null;
+      if (targetElement && !elementContains(root.current, targetElement)) {
+        focusFTZ();
+        internalState.hasFocus = true; // set focus here since we stop event propagation
+        ev.preventDefault();
+        ev.stopPropagation();
       }
-    },
-    [disabled, id, focus, internalState],
-  );
-
-  const updateEventHandlers = React.useCallback((): void => {
-    if (forceFocusInsideTrap && !internalState.disposeFocusHandler) {
-      internalState.disposeFocusHandler = on(window, 'focus', forceFocusInTrap, true);
-    } else if (!forceFocusInsideTrap && internalState.disposeFocusHandler) {
-      internalState.disposeFocusHandler();
-      internalState.disposeFocusHandler = undefined;
     }
+  });
 
-    if (!isClickableOutsideFocusTrap && !internalState.disposeClickHandler) {
-      internalState.disposeClickHandler = on(window, 'click', forceClickInTrap, true);
-    } else if (isClickableOutsideFocusTrap && internalState.disposeClickHandler) {
-      internalState.disposeClickHandler();
-      internalState.disposeClickHandler = undefined;
+  // Update window event handlers when relevant props change
+  React.useEffect(() => {
+    const disposables: Array<() => void> = [];
+
+    if (forceFocusInsideTrap) {
+      disposables.push(on(window, 'focus', forceFocusOrClickInTrap, true));
+    }
+    if (!isClickableOutsideFocusTrap) {
+      disposables.push(on(window, 'click', forceFocusOrClickInTrap, true));
     }
-  }, [forceClickInTrap, forceFocusInTrap, forceFocusInsideTrap, isClickableOutsideFocusTrap, internalState]);
 
-  // Updates eventHandlers and cleans up focusStack when the component unmounts.
-  React.useEffect(() => {
-    const parentRoot = root.current;
-    updateEventHandlers();
     return () => {
-      // don't handle return focus unless forceFocusInsideTrap is true or focus is still within FocusTrapZone
-      if (!disabled || forceFocusInsideTrap || !elementContains(parentRoot, doc?.activeElement as HTMLElement)) {
-        returnFocusToInitiator();
-      }
+      disposables.forEach(dispose => dispose());
     };
-    // eslint-disable-next-line react-hooks/exhaustive-deps -- Should only run on mount.
-  }, [updateEventHandlers]);
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- should only run when these two props change
+  }, [forceFocusInsideTrap, isClickableOutsideFocusTrap]);
 
-  // Updates focusStack and the previouslyFocusedElementOutsideTrapZone on prop change.
+  // On prop change or first render, focus the FTZ and update focusStack if appropriate
   React.useEffect(() => {
-    const newForceFocusInsideTrap = forceFocusInsideTrap !== undefined ? forceFocusInsideTrap : true;
```

Comment:
This effect has the most substantial bug related to prop transitions: it was an incorrect attempt to imitate the previous props/state behavior of `componentDidUpdate` (it's not actually saving the old values).

With the way that effects and their automatic cleanup works, it turns out that the only place we actually need to look at previous values is determining whether or not this is the first render (where we may need to initially force focus inside the trap).